runtime/src/account_storage.rs

```diff
@@ -211,6 +211,35 @@ impl AccountStorage {
     }
 }
 
+/// iterate contents of AccountStorage without exposing internals
+pub struct AccountStorageIter<'a> {
+    iter: dashmap::iter::Iter<'a, Slot, SlotStores>,
+}
+
+impl<'a> AccountStorageIter<'a> {
+    pub fn new(storage: &'a AccountStorage) -> Self {
+        Self {
+            iter: storage.map.iter(),
+        }
+    }
+}
+
+impl<'a> Iterator for AccountStorageIter<'a> {
+    type Item = (Slot, SnapshotStorageOne);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        for entry in self.iter.by_ref() {
+            // if no stores for a slot, then don't return the item at all, loops to try next slot
+            let slot = entry.key();
+            let stores = entry.value();
+            if let Some((_, store)) = stores.read().unwrap().iter().next() {
+                return Some((*slot, Arc::clone(store)));
```

Comment:
we can't return `&'a Arc<...>` because we got the arc from a read lock. `iter()` is not called often. Another way to do this is to pass a closure to the iter fn. Of the calls to `iter()`: One call is startup after index generation, 1 is debug, 1 is when we get the stores for a snapshot.

Reply:
> get_snapshot_storages()
An Arc is needed, but could be arguably simpler in the filter_map() by not wrapping the storage in an Option. (However, maybe the Option or the Arc is required here since there's a parallel-loop afterwards.
The `Option` was so that I could grab it and use it out without cloning it again in a par iter (so I could call `std::mem::take`). The `Arc` doesn't have a default value possible, so the option was to work around this. I couldn't find something like `into_par_chunks`. Mut was the best I could find.